File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3086_minimum-moves-to-pick-k-ones/solutions/cpp/solution.cpp
#include <vector>
#include <deque>
#include <algorithm>

using namespace std;

class Solution {
public:
    long long minimumMoves(vector<int>& nums, int k, int maxChanges) {
        int n = nums.size();
        vector<int> positions; // Stores the indices of '1' in nums
        
        // Collect positions of all ones in the array
        for (int i = 0; i < n; ++i) {
            if (nums[i] == 1) {
                positions.push_back(i);
            }
        }
        
        // We need to find the minimum number of moves to collect k ones
        long long result = LLONG_MAX;
        long long windowSum = 0; // sum of distances in the current window of k ones

        // We use a sliding window over the positions of the ones to find the optimal segment
        for (int i = 0; i < k; ++i) {
            // We consider the middle of the window as the initial pick up point
            windowSum += positions[i];
        }
        
        // The median index among the first k ones
        int medianIndex = k / 2;
        long long median = positions[medianIndex];
        // Calculate initial cost based on this median
        long long initialCost = 0;
        for (int i = 0; i < k; ++i) {
            initialCost += abs(positions[i] - median);
        }
        result = initialCost;
        
        // Sliding the window across all valid sets of k '1's in the array
        for (int i = k; i < positions.size(); ++i) {
            // Sliding the window: remove the leftmost one, add the new one
            windowSum += positions[i] - positions[i - k];
            // Recompute median position index and its value in the window
            medianIndex = (i - k + 1) + k / 2;
            median = positions[medianIndex];
            
            // Calculate the new cost
            long long newCost = 0;
            for (int j = i - k + 1; j <= i; ++j) {
                newCost += abs(positions[j] - median);
            }
            // Compare with the current best result
            result = min(result, newCost);
        }
        
        return result;
    }
};

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3086_minimum-moves-to-pick-k-ones/solutions/java/solution.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public long minimumMoves(int[] nums, int k, int maxChanges) {
        int n = nums.length;
        // Prefix sum array to calculate number of 1s in a given range quickly.
        int[] prefixSum = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            prefixSum[i] = prefixSum[i - 1] + nums[i - 1];
        }

        // Deque to store indices of 1s in the current sliding window.
        Deque<Integer> indicesOfOnes = new ArrayDeque<>();
        // Initialize the left pointer for the sliding window.
        int left = 0;
        long minMoves = Long.MAX_VALUE;

        for (int right = 0; right < n; right++) {
            if (nums[right] == 1) {
                indicesOfOnes.addLast(right);
            }
            if (indicesOfOnes.size() > k) {
                indicesOfOnes.removeFirst();
                left = indicesOfOnes.peekFirst();
            }
            if (indicesOfOnes.size() == k) {
                long currentMin = 0;
                // Calculate the median index of the 1s within the window
                int medianIndex = indicesOfOnes.peekFirst() + k / 2;
                int pos = 0;
                for (int idx : indicesOfOnes) {
                    currentMin += Math.abs(idx - medianIndex) - pos++;
                }
                // Calculate the number of ones and zeros in the left and right of medianIndex
                int onesLeft = prefixSum[medianIndex] - prefixSum[left];
                int zerosLeft = medianIndex - left - onesLeft;
                int onesRight = prefixSum[right + 1] - prefixSum[medianIndex + 1];
                int zerosRight = right - medianIndex - onesRight;
                // To minimize moves, calculate the necessary shifts within the given changes
                long neededMoves = currentMin + Math.min(
                    (long) Math.max(0, zerosLeft - maxChanges) + Math.max(0, zerosRight - maxChanges),
                    (long) Math.max(0, zerosLeft) + Math.max(0, zerosRight)
                );
                minMoves = Math.min(minMoves, neededMoves);
            }
        }

        return minMoves;
    }
}
